private/bufpkg/bufconnect/static_token_provider.go

```diff
@@ -45,55 +37,90 @@ func NewTokenProviderFromString(token string) (TokenProvider, error) {
-				return nil, fmt.Errorf("cannot parse token: two buf token provided: %q and %q", token, tokenProvider.defaultToken)
-			}
-			tokenProvider.defaultToken = token
+	if len(tokens) == 1 {
+		if strings.Contains(tokens[0], "@") {
+			return newMultipleTokenProvider(tokens, isFromEnvVar)
 		}
+		return newSingleTokenProvider(tokens[0], isFromEnvVar)
 	}
-	return tokenProvider, nil
+	return newMultipleTokenProvider(tokens, isFromEnvVar)
 }
 
-// RemoteToken finds the token by the remote address
-func (t *staticTokenProvider) RemoteToken(address string) string {
-	if authKeyPair, ok := t.tokenToAuthKey[address]; ok {
-		return authKeyPair.token
+// singleTokenProvider is used to provide set of authentication tokenToAuthKey.
+type singleTokenProvider struct {
+	// true: the tokenSet is generated from environment variable tokenEnvKey
+	// false: otherwise
+	setBufTokenEnvVar bool
+	token             string
+}
+
+func newSingleTokenProvider(token string, isFromEnvVar bool) (*singleTokenProvider, error) {
+	if strings.Contains(token, "@") {
+		return nil, errors.New("token cannot contain special character `@`")
+	}
+	if token == "" {
+		return nil, errors.New("single token cannot be empty")
 	}
-	return t.defaultToken
+	return &singleTokenProvider{
+		setBufTokenEnvVar: isFromEnvVar,
+		token:             token,
+	}, nil
 }
 
-func (t *staticTokenProvider) IsFromEnvVar() bool {
+// RemoteToken finds the token by the remote address
+func (t *singleTokenProvider) RemoteToken(address string) string {
+	return t.token
+}
+
+func (t *singleTokenProvider) IsFromEnvVar() bool {
 	return t.setBufTokenEnvVar
 }
 
-type authKey struct {
-	username string
-	token    string
+type multipleTokenProvider struct {
+	addressToToken map[string]string
+	isFromEnvVar   bool
 }
 
-func (ak *authKey) unmarshalString(s string) error {
-	username, token, found := strings.Cut(s, ":")
-	if !found {
-		return fmt.Errorf("cannot parse remote token: %s", s)
+func newMultipleTokenProvider(tokens []string, isFromEnvVar bool) (*multipleTokenProvider, error) {
+	addressToToken := make(map[string]string)
+	for _, token := range tokens {
+		split := strings.Split(token, "@")
+		if len(split) != 2 {
+			return nil, fmt.Errorf("invalid token: %s", token)
+		}
+		if split[0] == "" || split[1] == "" {
+			return nil, fmt.Errorf("invalid token: %s", token)
+		}
+		if _, ok := addressToToken[split[1]]; ok {
+			return nil, fmt.Errorf("invalid token: %s, repeated remote adddress: %s", token, split[1])
+		}
+		addressToToken[split[1]] = split[0]
 	}
-	ak.username = username
-	ak.token = token
-	return nil
+	return &multipleTokenProvider{
+		addressToToken: addressToToken,
+		isFromEnvVar:   isFromEnvVar,
+	}, nil
+}
+
+func (m *multipleTokenProvider) RemoteToken(address string) string {
+	return m.addressToToken[address]
+}
+
+func (m *multipleTokenProvider) IsFromEnvVar() bool {
+	return m.isFromEnvVar
+}
+
+type nonTokenProvider struct{}
+
+func (*nonTokenProvider) RemoteToken(string) string {
```

Comment:
nit: for nop structs, you can do the methods on the struct not the struct pointer, ie `func (nonTokenProvider)`, and return `nopTokenProvider` instead of `&nopTokenProvider`